File: engine.py
from __future__ import annotations

from dataclasses import dataclass, field
from enum import Enum
from heapq import heappush, heappop
from itertools import count
from typing import Any, Dict, List, Optional, Tuple

TREASURY_USER = "TREASURY" # special user for initial asset distribution and fees

class Side(str, Enum):
    BUY = "BUY"
    SELL = "SELL"

class OrderStatus(str, Enum):
    OPEN = "OPEN"
    FILLED = "FILLED"
    PARTIALLY_FILLED = "PARTIALLY_FILLED"
    CANCELED = "CANCELED"
    REJECTED = "REJECTED"

_order_id_gen = count(1)
_trade_id_gen = count(1)
_seq_gen = count(1) # Global time-priority sequence generator for all events
_event_id_gen = count(1) # Global event ID generator for all events (orders, trades, cancels, etc)

@dataclass(frozen=True)
class NewOrder:
    user_id: str
    asset_id: str
    side: Side
    qty: int
    limit_price_cents: int # store money as int cents
    client_order_id: Optional[str] = None

@dataclass
class Order:
    id: int
    user_id: str
    asset_id: str
    side: Side
    qty: int
    remaining_qty: int
    limit_price_cents: int
    status: OrderStatus
    seq: int    # time priority (lower = higher priority)

@dataclass(frozen=True)
class Trade:
    id: int
    asset_id: str
    price_cents: int
    qty: int
    buy_order_id: int
    sell_order_id: int
    buyer_id: str
    seller_id: str

@dataclass
class Account:
    cash_cents: int
    reserved_cash_cents: int = 0

@dataclass
class Holding:
    shares: int
    reserved_shares: int = 0

@dataclass(frozen=True)
class Asset:
    asset_id: str
    issuer_user_id: str
    total_supply: int
    name: str

class EventType(str, Enum):
    ASSET_CREATED = "ASSET_CREATED"
    ORDER_PLACED = "ORDER_PLACED"
    ORDER_CANCELED = "ORDER_CANCELED"
    TRADE_EXECUTED = "TRADE_EXECUTED"
    CASH_MOVED = "CASH_MOVED"
    SHARES_MOVED = "SHARES_MOVED"

@dataclass(frozen=True)
class Event:
    id: int
    type: EventType
    ts_seq: int  # reuses global _seq_gen for ordering
    data: Dict[str, Any] = field(default_factory=dict)

class InsufficientFunds(Exception):
    pass

class InsufficientShares(Exception):
    pass

class UnknownOrder(Exception):
    pass

class Book:
    """
    Priority queue for an asset:
        - bids: max price, then earlier time (use negative price)
        - asks: min price, then earlier time
    tuples are: (key_price, seq, order_id)
    """
    def __init__(self) -> None:
        self.bids: List[Tuple[int, int, int]] = []
        self.asks: List[Tuple[int, int, int]] = []

    
class MatchingEngine:
    def __init__(self) -> None:
        self.accounts: Dict[str, Account] = {}
        self.holdings: Dict[Tuple[str, str], Holding] = {}  # (user_id, asset_id)
        self.orders: Dict[int, Order] = {}
        self.books: Dict[str, Book] = {} # asset_id -> Book
        self.last_price_cents: Dict[str, int] = {} # asset_id -> last trade price
        self.assets: Dict[str, Asset] = {}
        self.events: List[Event] = [] # global event log for all state changes (orders, trades, cancels, etc)

    # ------ Public API ------

    def ensure_user(self, user_id: str, initial_cash_cents: int = 0) -> None:
        self.accounts.setdefault(user_id, Account(cash_cents=initial_cash_cents))

    def ensure_asset(self, asset_id: str, initial_price_cents: int = 1000) -> None:
        self.books.setdefault(asset_id, Book())
        self.last_price_cents.setdefault(asset_id, initial_price_cents)

    def create_person_asset(
            self, 
            issuer_user_id: str, 
            asset_id: str,
            total_supply: int = 1000,
            issuer_pct: float = 0.6
            ) -> None:
        
        if asset_id in self.assets:
            raise ValueError("Asset already exists")
        if total_supply <= 0:
            raise ValueError("Total supply must be positive")
        if not (0 < issuer_pct < 1):
            raise ValueError("Issuer percentage must be between 0 and 1")

        self.assets[asset_id] = Asset(
            asset_id=asset_id,
            issuer_user_id=issuer_user_id,
            total_supply=total_supply,
            # TODO - allow custom names
            name=f"USER{asset_id}"
        )

        issuer_shares = int(round(total_supply * issuer_pct))
        treasury_shares = total_supply - issuer_shares

        self._getholding(issuer_user_id, asset_id).shares += issuer_shares
        self._getholding(TREASURY_USER, asset_id).shares += treasury_shares

        # Emit asset creation event with initial distribution details
        seq = next(_seq_gen)
        self._emit(
            EventType.ASSET_CREATED, 
            seq, 
            asset_id=asset_id, 
            issuer_user_id=issuer_user_id, 
            total_supply=total_supply, 
            distribution=[
                {"user_id": issuer_user_id, "shares": issuer_shares},         
                {"user_id": TREASURY_USER, "shares": treasury_shares},
            ],
        )

        self._emit(EventType.SHARES_MOVED, next(_seq_gen), asset_id=asset_id, from_user_id=None, to_user_id=issuer_user_id, shares=issuer_shares, reason="ISSUANCE")
        self._emit(EventType.SHARES_MOVED, next(_seq_gen), asset_id=asset_id, from_user_id=None, to_user_id="TREASURY", shares=treasury_shares, reason="ISSUANCE")


    def place_order(self, req: NewOrder) -> Tuple[Order, List[Trade]]:
        self._validate_new_order(req)

        order = Order(
            id=next(_order_id_gen),
            user_id=req.user_id,
            asset_id=req.asset_id,
            side=req.side,
            qty=req.qty,
            remaining_qty=req.qty,
            limit_price_cents=req.limit_price_cents,
            status=OrderStatus.OPEN,
            seq=next(_seq_gen)
        )

        # Reserve upfront so we never go negative during matching
        self._reserve_for_order(order)

        seq = order.seq
        self._emit(
            EventType.ORDER_PLACED, 
            seq, 
            order_id=order.id, 
            user_id=order.user_id, 
            asset_id=order.asset_id, 
            side=order.side, 
            qty=order.qty, 
            limit_price_cents=order.limit_price_cents
        )

        # Match immediately against the opposite side
        trades: List[Trade] = self._match(order)

        # If still remaining, rest on the book
        if order.remaining_qty > 0 and order.status not in (OrderStatus.REJECTED, OrderStatus.CANCELED):
            self._add_to_book(order)
            if order.remaining_qty < order.qty:
                order.status = OrderStatus.PARTIALLY_FILLED

        else:
            order.status = OrderStatus.FILLED if order.remaining_qty == 0 else order.status

        self.orders[order.id] = order
        return order, trades

    def cancel_order(self, order_id: int) -> Order:
        """
        Simple cancel: mark status and release remaining reserves.
        Notes: For a production engine we need to remove from it the heap efficiently
        For MVP, we do lazy deletion (skip non-open orders when popping).
        """
        order = self.orders.get(order_id)
        if order is None:
            raise UnknownOrder(order_id)
        if order.status in (OrderStatus.FILLED, OrderStatus.CANCELED, OrderStatus.REJECTED):
            return order
        
        order.status = OrderStatus.CANCELED
        self._release_remaining_reserve(order)

        seq = next(_seq_gen)
        self._emit(
            EventType.ORDER_CANCELED, 
            seq, 
            order_id=order.id, 
            user_id=order.user_id, 
            asset_id=order.asset_id, 
            side=order.side, 
            remaining_qty=order.remaining_qty
        )
        return order
    
    # ------ Matching Logic ------

    def _match(self, incoming: Order) -> List[Trade]:
        book = self.books[incoming.asset_id]
        trades: List[Trade] = []

        while incoming.remaining_qty > 0:
            # peek automatically discards non-open orders, so we are guaranteed a valid resting order or None
            best_order_id = self._peek_best_opposite(book, incoming.side)
            if best_order_id is None:
                break

            resting = self.orders[best_order_id]

            if not self._is_price_cross(incoming, resting):
                break

            # Pop it now to trade against it
            self._pop_best_opposite(book, incoming.side)

            fill_qty = min(incoming.remaining_qty, resting.remaining_qty)
            trade_price = resting.limit_price_cents # price is determined by resting order

            trade = self._execute_trade(
                asset_id=incoming.asset_id,
                price_cents=trade_price,
                qty=fill_qty,
                taker=incoming,
                maker=resting
            )
            trades.append(trade)

            # If the resting order still has remaining qty, put it back on the book
            if resting.remaining_qty > 0:
                resting.status = OrderStatus.PARTIALLY_FILLED
                self._add_to_book(resting)
            else:
                resting.status = OrderStatus.FILLED
            
            if incoming.remaining_qty == 0:
                incoming.status = OrderStatus.FILLED
            
        return trades
        
    def _execute_trade(self, asset_id: str, price_cents: int, qty: int, taker: Order, maker: Order):
        # determine buyer/seller based on sides
        if taker.side == Side.BUY:
            buyer, seller = taker, maker
        else:
            buyer, seller = maker, taker
        
        notional = price_cents * qty
        
        # Buyer: reduce reserved at LIMIT*qty, spend at TRADE*qty
        self._consume_buy_reserve_on_fill(buyer, qty)
        self.accounts[buyer.user_id].cash_cents -= notional

        # Seller: reduce reserved shares, receive cash
        self._consume_shares_reserve(seller.user_id, asset_id, qty)
        self._getholding(seller.user_id, asset_id).shares -= qty
        self.accounts[seller.user_id].cash_cents += notional

        # Buyer receives shares
        self._getholding(buyer.user_id, asset_id).shares += qty

        # Update order remaining
        taker.remaining_qty -= qty
        maker.remaining_qty -= qty

        # Last price
        self.last_price_cents[asset_id] = price_cents

        seq = next(_seq_gen)
        trade = Trade(
            id=next(_trade_id_gen),
            asset_id=asset_id,
            price_cents=price_cents,
            qty=qty,
            buy_order_id=buyer.id,
            sell_order_id=seller.id,
            buyer_id=buyer.user_id,
            seller_id=seller.user_id
        )

        # Emit trade event with details as well as account movements
        self._emit(
            EventType.TRADE_EXECUTED,
            seq,
            trade_id=trade.id,
            asset_id=asset_id,
            price_cents=price_cents,
            qty=qty,
            buy_order_id=buyer.id,
            sell_order_id=seller.id,
            buyer_id=buyer.user_id,
            seller_id=seller.user_id
        )

        self._emit(EventType.CASH_MOVED, next(_seq_gen), asset_id=asset_id, from_user_id=buyer.user_id, to_user_id=seller.user_id, cash_cents=notional, reason="TRADE")
        self._emit(EventType.SHARES_MOVED, next(_seq_gen), asset_id=asset_id, from_user_id=seller.user_id, to_user_id=buyer.user_id, shares=qty, reason="TRADE")

        return trade
    
    # ------ Book Helpers ------

    def _add_to_book(self, order: Order) -> None:
        book = self.books[order.asset_id]
        if order.side == Side.BUY:
            # max-heap by price using negative price, then time priority
            heappush(book.bids, (-order.limit_price_cents, order.seq, order.id))
        else:
            heappush(book.asks, (order.limit_price_cents, order.seq, order.id))

        self.orders[order.id] = order # ensure present

    def _peek_best_opposite(self, book: Book, incoming_side: Side) -> Optional[int]:
        heap = book.asks if incoming_side == Side.BUY else book.bids
        while heap:
            _, _, order_id = heap[0]
            o = self.orders.get(order_id)
            if o is not None and self._is_open(o):
                return order_id
            heappop(heap) # discard lazy-dead top
        return None   

    def _pop_best_opposite(self, book: Book, incoming_side: Side) -> Optional[int]:
        heap = book.asks if incoming_side == Side.BUY else book.bids
        while heap:
            _, _, order_id = heappop(heap)
            o = self.orders.get(order_id)
            if o is not None and self._is_open(o):
                return order_id
        return None
    
    def _is_price_cross(self, incoming: Order, resting: Order) -> bool:
        if incoming.side == Side.BUY:
            return incoming.limit_price_cents >= resting.limit_price_cents
        else:
            return incoming.limit_price_cents <= resting.limit_price_cents
        
    def _is_open(self, order: Order) -> bool:
        return order.status in [OrderStatus.OPEN, OrderStatus.PARTIALLY_FILLED] and order.remaining_qty > 0
    
    # ------ Validation and reserves ------
    def _validate_new_order(self, req: NewOrder) -> None:
        if req.qty <= 0:
            raise ValueError("Order quantity must be positive")
        if req.limit_price_cents <= 0:
            raise ValueError("Limit price must be positive")
        self.ensure_user(req.user_id)
        self.ensure_asset(req.asset_id)

    def _reserve_for_order(self, order: Order) -> None:
        if order.side == Side.BUY:
            needed = order.limit_price_cents * order.remaining_qty
            acct = self.accounts[order.user_id]
            available = acct.cash_cents - acct.reserved_cash_cents
            if available < needed:
                raise InsufficientFunds(f"need: {needed}, available: {available}")
            acct.reserved_cash_cents += needed
        else:
            h = self._getholding(order.user_id, order.asset_id)
            available = h.shares - h.reserved_shares
            if available < order.remaining_qty:
                raise InsufficientShares(f"need: {order.remaining_qty}, available: {available}")
            h.reserved_shares += order.remaining_qty

    def _consume_buy_reserve_on_fill(self, buy_order: Order, fill_qty: int) -> None:
        acct = self.accounts[buy_order.user_id]
        release = buy_order.limit_price_cents * fill_qty
        if acct.reserved_cash_cents < release:
            raise InsufficientFunds("reserved cash underflow on fill")
        acct.reserved_cash_cents -= release

    def _consume_shares_reserve(self, user_id: str, asset_id: str, shares: int) -> None:
        h = self._getholding(user_id, asset_id)
        if h.reserved_shares < shares:
            raise InsufficientShares("reserved shares underflow")
        h.reserved_shares -= shares
    
    def _release_remaining_reserve(self, order: Order) -> None:
        if order.remaining_qty <= 0:
            return
        if order.side == Side.BUY:
            refund = order.limit_price_cents * order.remaining_qty
            acct = self.accounts[order.user_id]
            acct.reserved_cash_cents = max(0, acct.reserved_cash_cents - refund)
        else:
            h = self._getholding(order.user_id, order.asset_id)
            h.reserved_shares = max(0, h.reserved_shares - order.remaining_qty)

        order.remaining_qty = 0

    # ------ State Helpers ------
    def _getholding(self, user_id: str, asset_id: str) -> Holding:
        key = (user_id, asset_id)
        if key not in self.holdings:
            self.holdings[key] = Holding(shares=0)
        return self.holdings[key]
    
    # ------ Event Logging ------
    def _emit(self, etype: EventType, ts_seq: int, **data: Any) -> Event:
        event = Event(
            id=next(_event_id_gen),
            type=etype,
            ts_seq=ts_seq,
            data=data
        )
        self.events.append(event)
        return event
    
    def rebuild_from_events(self) -> None:
        # wipe derived state
        for acct in self.accounts.values():
            acct.cash_cents = 0
            acct.reserved_cash_cents = 0
        for h in self.holdings.values():
            h.shares = 0
            h.reserved_shares = 0
        
        # for now just rebuild balances from ledger events
        for event in sorted(self.events, key=lambda e: (e.ts_seq, e.id)):
            if event.type == EventType.CASH_MOVED:
                from_user = event.data["from_user_id"]
                to_user = event.data["to_user_id"]
                cash_cents = event.data["cash_cents"]
                if from_user:
                    self.accounts[from_user].cash_cents -= cash_cents
                if to_user:
                    self.accounts[to_user].cash_cents += cash_cents
                    
            elif event.type == EventType.SHARES_MOVED:
                from_user = event.data["from_user_id"]
                to_user = event.data["to_user_id"]
                asset_id = event.data["asset_id"]
                shares = event.data["shares"]
                if from_user:
                    self._getholding(from_user, asset_id).shares -= shares
                if to_user:
                    self._getholding(to_user, asset_id).shares += shares


# ------ Usage Example ------
if __name__ == "__main__":
    eng = MatchingEngine()
    eng.ensure_user("alice", initial_cash_cents=100_000)  # $1000.00
    eng.ensure_user("bob", initial_cash_cents=100_000)
    eng.ensure_user(TREASURY_USER)

    eng.ensure_asset("bob-stock", initial_price_cents=1_000)
    eng.create_person_asset("bob-stock", "bob")

    # Bob posts a sell: 10 shares at $12
    o1, t1 = eng.place_order(NewOrder(
        user_id="bob", asset_id="bob-stock", side=Side.SELL, qty=10, limit_price_cents=1_200
    ))

    # Alice buys: 7 shares at $15 (crosses, fills at $12)
    o2, t2 = eng.place_order(NewOrder(
        user_id="alice", asset_id="bob-stock", side=Side.BUY, qty=7, limit_price_cents=1_500
    ))

    print(o1)
    print(o2)
    print(t2)
    print("Alice cash:", eng.accounts["alice"].cash_cents, "shares:", eng.holdings[("alice", "bob-stock")].shares)
    print("Bob cash:", eng.accounts["bob"].cash_cents, "shares:", eng.holdings[("bob", "bob-stock")].shares)
        

File: server/src/engine/engine.js
import {
    TREASURY_USER,
    Side,
    OrderStatus,
    EventType,
    InsufficientShares,
    InsufficientFunds,
    UnknownOrder,
    UnknownUser,
    UnknownAsset,
} from "./types.js";
import { Heap } from "./heaps.js";

function makeIdGen(start = 1) {
    let x = start;
    return () => x++;
}

export function makeEngine() {
    return new MatchingEngine();
}

class Book {
  constructor() {
    // Each heap entry is { price, seq, orderId }
    // bids: higher price first, then lower seq
    this.bids = new Heap((a, b) => {
      if (a.price !== b.price) return a.price > b.price;
      return a.seq < b.seq;
    });

    // asks: lower price first, then lower seq
    this.asks = new Heap((a, b) => {
      if (a.price !== b.price) return a.price < b.price;
      return a.seq < b.seq;
    });
  }
}

class MatchingEngine {
    constructor() {
    this._orderId = makeIdGen(1);
    this._tradeId = makeIdGen(1);
    this._eventId = makeIdGen(1);
    this._seq = makeIdGen(1);

    this.accounts = new Map();  // userId -> { cash, reservedCash }
    this.holdings = new Map();  // `${userId}|${assetId}` -> { shares, reservedShares }
    this.assets = new Map();    // assetId -> { assetId, issuerUserId, totalSupply, name }
    this.orders = new Map();    // orderId -> order
    this.books = new Map();     // assetId -> Book
    this.lastPrice = new Map(); // assetId -> last price cents

    this.events = [];           // append-only
  }

  // ---------- Setup ----------
  setUserDefault(userId, initialCashCents = 0) {
    if (!this.accounts.has(userId)) {
      this.accounts.set(userId, { cash: initialCashCents, reservedCash: 0 });
    } else {
      const acct = this.accounts.get(userId);
      acct.cash = initialCashCents;
      acct.reservedCash = 0;
    }
  }

  ensureUser(userId) {
    if (!this.accounts.has(userId)) {
      this.accounts.set(userId, { cash: 0, reservedCash: 0 });
    }
  }

  ensureBook(assetId, initialPriceCents = 1000) {
    if (!this.books.has(assetId)) this.books.set(assetId, new Book());
    if (!this.lastPrice.has(assetId)) this.lastPrice.set(assetId, initialPriceCents);
  }

  validateUser(userId) {
    if (!this.accounts.has(userId)) throw new UnknownUser(userId);
  }

  validateAsset(assetId) {
    if (!this.assets.has(assetId)) throw new UnknownAsset(assetId);
  }

  createPersonAsset({ issuerUserId, assetId, totalSupply = 1000, issuerPct = 0.6, name = null }) {
    if (this.assets.has(assetId)) throw new Error("asset already exists");
    if (totalSupply <= 0) throw new Error("totalSupply must be > 0");
    if (!(issuerPct > 0 && issuerPct < 1)) throw new Error("issuerPct must be between 0 and 1");

    this.ensureUser(issuerUserId);
    this.ensureUser(TREASURY_USER);
    this.ensureBook(assetId);

    const issuerShares = Math.round(totalSupply * issuerPct);
    const treasuryShares = totalSupply - issuerShares;

    this.assets.set(assetId, {
      assetId,
      issuerUserId,
      totalSupply,
      name: name ?? `${issuerUserId}'s ${assetId}`,
    });

    this._getHolding(issuerUserId, assetId).shares += issuerShares;
    this._getHolding(TREASURY_USER, assetId).shares += treasuryShares;

    const seq = this._nextSeq();
    this._emit(EventType.ASSET_CREATED, seq, {
      asset_id: assetId,
      issuer_user_id: issuerUserId,
      total_supply: totalSupply,
      distribution: [
        { user_id: issuerUserId, shares: issuerShares },
        { user_id: TREASURY_USER, shares: treasuryShares },
      ],
    });

    this._emit(EventType.SHARES_MOVED, this._nextSeq(), {
      asset_id: assetId,
      from_user_id: null,
      to_user_id: issuerUserId,
      shares: issuerShares,
      reason: "ISSUANCE",
    });

    if (treasuryShares !== 0) {
      this._emit(EventType.SHARES_MOVED, this._nextSeq(), {
        asset_id: assetId,
        from_user_id: null,
        to_user_id: TREASURY_USER,
        shares: treasuryShares,
        reason: "ISSUANCE",
      });
    }

    return { events: this._drainNewEvents() };
  }

  // ---------- Trading API ----------
  processOrder(req) {
    this._validateNewOrder(req);

    const order = {
      id: this._orderId(),
      userId: req.userId,
      assetId: req.assetId,
      side: req.side,
      qty: req.qty,
      remainingQty: req.qty,
      limitPriceCents: req.limitPriceCents,
      status: OrderStatus.OPEN,
      seq: this._nextSeq(),
    };

    // reserve upfront
    this._reserveForOrder(order);

    this._emit(EventType.ORDER_PLACED, order.seq, {
      order_id: order.id,
      user_id: order.userId,
      asset_id: order.assetId,
      side: order.side,
      qty: order.qty,
      limit_price_cents: order.limitPriceCents,
    });

    // match
    const trades = this._match(order);

    // rest remaining
    if (order.remainingQty > 0 && order.status !== OrderStatus.CANCELED && order.status !== OrderStatus.REJECTED) {
      this._addToBook(order);
      if (order.remainingQty < order.qty) order.status = OrderStatus.PARTIALLY_FILLED;
    } else {
      if (order.remainingQty === 0) order.status = OrderStatus.FILLED;
    }

    this.orders.set(order.id, order);

    return { order, trades, events: this._drainNewEvents() };
  }

  cancelOrder(orderId) {
    const order = this.orders.get(orderId);
    if (!order) throw new UnknownOrder(orderId);
    if ([OrderStatus.FILLED, OrderStatus.CANCELED, OrderStatus.REJECTED].includes(order.status)) {
      return { order, events: [] };
    }

    order.status = OrderStatus.CANCELED;
    this._releaseRemainingReserve(order);

    const seq = this._nextSeq();
    this._emit(EventType.ORDER_CANCELED, seq, {
      order_id: order.id,
      user_id: order.userId,
      asset_id: order.assetId,
      side: order.side,
      remaining_qty: 0,
    });

    return { order, events: this._drainNewEvents() };
  }

  _match(incoming) {
    this.ensureBook(incoming.assetId);
    const book = this.books.get(incoming.assetId);
    const trades = [];

    while (incoming.remainingQty > 0) {
      const bestId = this._peekBestOpposite(book, incoming.side);
      if (bestId == null) break;

      const resting = this.orders.get(bestId);
      if (!resting || !this._isOpen(resting)) {
        this._popBestOpposite(book, incoming.side);
        continue;
      }

      if (!this._isPriceCross(incoming, resting)) break;

      this._popBestOpposite(book, incoming.side);

      const fillQty = Math.min(incoming.remainingQty, resting.remainingQty);
      const tradePrice = resting.limitPriceCents;

      const trade = this._executeTrade({
        assetId: incoming.assetId,
        priceCents: tradePrice,
        qty: fillQty,
        taker: incoming,
        maker: resting,
      });

      trades.push(trade);

      if (resting.remainingQty > 0) {
        resting.status = OrderStatus.PARTIALLY_FILLED;
        this._addToBook(resting);
      } else {
        resting.status = OrderStatus.FILLED;
      }

      if (incoming.remainingQty === 0) incoming.status = OrderStatus.FILLED;
    }

    return trades;
  }

   _executeTrade({ assetId, priceCents, qty, taker, maker }) {
    const isTakerBuy = taker.side === Side.BUY;
    const buyer = isTakerBuy ? taker : maker;
    const seller = isTakerBuy ? maker : taker;

    const notional = priceCents * qty;

    // Buy-side reserve release at LIMIT*qty, spend at TRADE*qty
    this._consumeBuyReserveOnFill(buyer, qty);
    this._acct(buyer.userId).cash -= notional;

    // Seller reserve shares, deliver shares, receive cash
    this._consumeSharesReserve(seller.userId, assetId, qty);
    this._getHolding(seller.userId, assetId).shares -= qty;
    this._acct(seller.userId).cash += notional;

    // Buyer receives shares
    this._getHolding(buyer.userId, assetId).shares += qty;

    taker.remainingQty -= qty;
    maker.remainingQty -= qty;

    this.lastPrice.set(assetId, priceCents);

    const trade = {
      id: this._tradeId(),
      assetId,
      priceCents,
      qty,
      buyOrderId: buyer.id,
      sellOrderId: seller.id,
      buyerId: buyer.userId,
      sellerId: seller.userId,
      tsSeq: this._nextSeq(),
    };

    this._emit(EventType.TRADE_EXECUTED, trade.tsSeq, {
      trade_id: trade.id,
      asset_id: assetId,
      price_cents: priceCents,
      qty,
      buy_order_id: trade.buyOrderId,
      sell_order_id: trade.sellOrderId,
      buyer_id: trade.buyerId,
      seller_id: trade.sellerId,
    });

    this._emit(EventType.CASH_MOVED, this._nextSeq(), {
      asset_id: assetId,
      from_user_id: trade.buyerId,
      to_user_id: trade.sellerId,
      cash_cents: notional,
      reason: "TRADE",
      trade_id: trade.id,
    });

    this._emit(EventType.SHARES_MOVED, this._nextSeq(), {
      asset_id: assetId,
      from_user_id: trade.sellerId,
      to_user_id: trade.buyerId,
      shares: qty,
      reason: "TRADE",
      trade_id: trade.id,
    });

    return trade;
  }

  // ---------- Book helpers ----------
  _addToBook(order) {
    this.ensureBook(order.assetId);
    const book = this.books.get(order.assetId);

    if (order.side === Side.BUY) {
      book.bids.push({ price: order.limitPriceCents, seq: order.seq, orderId: order.id });
    } else {
      book.asks.push({ price: order.limitPriceCents, seq: order.seq, orderId: order.id });
    }

    this.orders.set(order.id, order);
  }

  _peekBestOpposite(book, incomingSide) {
    const heap = incomingSide === Side.BUY ? book.asks : book.bids;
    while (heap.size()) {
      const top = heap.peek();
      const o = this.orders.get(top.orderId);
      if (o && this._isOpen(o)) return top.orderId;
      heap.pop(); // lazy delete
    }
    return null;
  }

  _popBestOpposite(book, incomingSide) {
    const heap = incomingSide === Side.BUY ? book.asks : book.bids;
    while (heap.size()) {
      const top = heap.pop();
      const o = this.orders.get(top.orderId);
      if (o && this._isOpen(o)) return top.orderId;
    }
    return null;
  }

  _isPriceCross(incoming, resting) {
    if (incoming.side === Side.BUY) return incoming.limitPriceCents >= resting.limitPriceCents;
    return incoming.limitPriceCents <= resting.limitPriceCents;
  }

  _isOpen(order) {
    return (order.status === OrderStatus.OPEN || order.status === OrderStatus.PARTIALLY_FILLED) && order.remainingQty > 0;
  }

  // ---------- Validation + reserves ----------
  _validateNewOrder(req) {
    if (!req || typeof req !== "object") throw new Error("bad order");
    if (req.qty <= 0) throw new Error("qty must be > 0");
    if (req.limitPriceCents <= 0) throw new Error("limitPriceCents must be > 0");
    this.validateUser(req.userId);
    this.validateAsset(req.assetId);
  }

  _reserveForOrder(order) {
    if (order.side === Side.BUY) {
      const needed = order.limitPriceCents * order.remainingQty;
      const acct = this._acct(order.userId);
      const available = acct.cash - acct.reservedCash;
      if (available < needed) {
        order.status = OrderStatus.REJECTED;
        throw new InsufficientFunds(`need ${needed}, available ${available}`);
      }
      acct.reservedCash += needed;
    } else {
      const h = this._getHolding(order.userId, order.assetId);
      const available = h.shares - h.reservedShares;
      if (available < order.remainingQty) {
        order.status = OrderStatus.REJECTED;
        throw new InsufficientShares(`need ${order.remainingQty}, available ${available}`);
      }
      h.reservedShares += order.remainingQty;
    }
  }

  _consumeBuyReserveOnFill(buyOrder, fillQty) {
    const acct = this._acct(buyOrder.userId);
    const release = buyOrder.limitPriceCents * fillQty;
    if (acct.reservedCash < release) throw new InsufficientFunds("reserved cash underflow on fill");
    acct.reservedCash -= release;
  }

  _consumeSharesReserve(userId, assetId, shares) {
    const h = this._getHolding(userId, assetId);
    if (h.reservedShares < shares) throw new InsufficientShares("reserved shares underflow");
    h.reservedShares -= shares;
  }

  _releaseRemainingReserve(order) {
    if (order.remainingQty <= 0) return;

    if (order.side === Side.BUY) {
      const refund = order.limitPriceCents * order.remainingQty;
      const acct = this._acct(order.userId);
      acct.reservedCash = Math.max(0, acct.reservedCash - refund);
    } else {
      const h = this._getHolding(order.userId, order.assetId);
      h.reservedShares = Math.max(0, h.reservedShares - order.remainingQty);
    }
    order.remainingQty = 0;
  }

  // ---------- Events + replay ----------
  _emit(type, tsSeq, data) {
    const ev = { id: this._eventId(), type, tsSeq, data };
    this.events.push(ev);
    return ev;
  }

  _drainNewEvents() {
    const out = this.events;
    this.events = [];
    return out;
  }

  rebuildFromEvents(rows) {
    // rows: [{ ts_seq, type, data }]
    // wipe balances/holdings
    for (const acct of this.accounts.values()) {
      acct.cash = 0;
      acct.reservedCash = 0;
    }
    for (const h of this.holdings.values()) {
      h.shares = 0;
      h.reservedShares = 0;
    }

    const sorted = [...rows].sort((a, b) => (a.ts_seq - b.ts_seq));
    for (const r of sorted) {
      if (r.type === EventType.CASH_MOVED) {
        const { from_user_id, to_user_id, cash_cents } = r.data;
        if (from_user_id) this._acct(from_user_id).cash -= cash_cents;
        if (to_user_id) this._acct(to_user_id).cash += cash_cents;
      }
      if (r.type === EventType.SHARES_MOVED) {
        const { asset_id, from_user_id, to_user_id, shares } = r.data;
        this.ensureBook(asset_id);
        if (from_user_id) this._getHolding(from_user_id, asset_id).shares -= shares;
        if (to_user_id) this._getHolding(to_user_id, asset_id).shares += shares;
      }
    }
  }

  // ---------- State helpers ----------
  _acct(userId) {
    this.ensureUser(userId);
    return this.accounts.get(userId);
  }

  _holdingKey(userId, assetId) {
    return `${userId}|${assetId}`;
  }

  _getHolding(userId, assetId) {
    const k = this._holdingKey(userId, assetId);
    if (!this.holdings.has(k)) this.holdings.set(k, { shares: 0, reservedShares: 0 });
    return this.holdings.get(k);
  }

  _nextSeq() {
    return this._seq();
  }
}